### ISSUE TITLE
Training details and VGG19_normalized

### ISSUE THREAD

**Issue by irasin** [black-puppydog/PytorchWCT#2]
Hi, thank you for your work.

I am implementing this work use pure Pytorch. So I want to know if I can find pre-trained VGG19_normalized model in Pytorch or how should I use it from other frameworks.

I also want to know what you mean about    

```python
#self.preprocess_weight =
    self.register_buffer(
      'preprocess_weight',
      torch.FloatTensor([[[[  0.]], [[  0.]], [[255.]]],
                         [[[  0.]], [[255.]], [[  0.]]],
                         [[[255.]], [[  0.]], [[  0.]]]]))
    #self.preprocess_bias =
    self.register_buffer(
      'preprocess_bias',
      torch.FloatTensor([-103.9390, -116.7790, -123.6800]))
```

in  vgg19_normalized.py

Thanks a lot.

**Comment by black-puppydog** [comment]
Hi @irasin,

no, from what I can tell, normalized vgg has never been and will likely not be any time soon in torchvision.
Leon Gatys himself actually published a set of weights he uses for his code:
https://github.com/leongatys/PytorchNeuralStyleTransfer/blob/master/download_models.sh

I use the weights I adapted from @sunshineatnoon's code. Those are lua models, that are probably actually adapted from caffe models etc etc :stuck_out_tongue: 

The code you quote takes care of three things:

1. It switches the color order of the input from RGB to BGR, because that's what the network expects. That's why the values are on the counter diagonal of the conv weight matrix.
2. It scales the input to the input range the network expects. Pytorch image loading puts the images into the range `[0, 1]` but this network expects image loading to be `uint8`, so in `[0, 255]`.
3. It removes the standard VGG mean activations from the image via the `bias`.

I use `register_buffer()` instead of `register_parameter()` because I do want these weights to be persisted with `torch.save(net.model_dict())` so that it is easy to later load the model simply with `torch.load()` without getting into trouble with missing parameters or such. But I do *not* want them to be considered as model parameters in the strict sense. From the pytorch docs:

> This is typically used to register a buffer that should not to be considered a model parameter. For example, BatchNorm’s running_mean is not a parameter, but is part of the persistent state.

I figured this is a good fit.
Let me know if there's anything else I can help with. :)

**Comment by irasin** [comment]
Hi, @black-puppydog, thank you for your reply.

I read @sunshineatnoon 's implementation and find that he didn't use the BGR order.

I used the pre-trained models from  https://github.com/pietrocarbo/deep-transfer and I think the effects of the output are not satisfying to me.

Here is my implementation.
https://github.com/irasin/Pytorch_WCT

I am a newbie in Pytorch so if you could give me some advice on how to improve the output, I will be very appreciated.

Thanks a lot a lot!!

**Comment by irasin** [comment]
Thanks a lot for your help.
